### ISSUE TITLE
Reproducibility fails when using MixUp

### ISSUE THREAD

**Issue by dpaleka** [mosaicml/composer#793]
## Environment

- OS: [Ubuntu 20.04]
- Hardware (GPU, or instance type): [GeForce GTX 1080 Ti]
- Relevant package versions:
```
mosaicml==0.5.0
numpy==1.21.5
torch==1.10.1
torchmetrics==0.7.0
torchvision==0.11.2
```

## Steps to reproduce the behavior:

1. Run the following code twice:
```
from torch.utils.data import DataLoader
from torchvision import datasets, transforms

import composer
from composer.algorithms import MixUp

composer.utils.reproducibility.configure_deterministic_mode()
composer.utils.reproducibility.seed_all(42)

mean = [0.507, 0.487, 0.441]
std  = [0.267, 0.256, 0.276]
cifar10_transforms = transforms.Compose([transforms.ToTensor(), transforms.Normalize(mean, std)])

train_dataset = datasets.CIFAR10("data", download=True, train=True, transform=cifar10_transforms)
eval_dataset = datasets.CIFAR10("data", download=True, train=False, transform=cifar10_transforms)
train_dataloader = DataLoader(train_dataset, batch_size=512, pin_memory=True, shuffle=False)
eval_dataloader = DataLoader(eval_dataset, batch_size=512, pin_memory=True, shuffle=False)

model = composer.models.CIFAR10_ResNet56(num_classes=10)

algorithms = [
    MixUp(num_classes=10, alpha=0.1)
]

trainer = composer.Trainer(
    device='gpu',
    model=model,
    train_dataloader=train_dataloader,
    eval_dataloader=eval_dataloader,
    max_duration="5ep",
    algorithms=algorithms,
) 

trainer.fit()
```
2. (Optional) Add  `print(f'mixing_lambda: {mixing_lambda}')` under the following line in the codebase, and repeat step 1: https://github.com/mosaicml/composer/blob/9387d5c6d9f22ee4e6b4fc0696227f4d84d96f28/composer/algorithms/mixup/mixup.py#L182

## Expected behavior
All runs should produce the same output.

## Additional context
This might be connected with seeding the numpy global random state, which is legacy since Numpy 1.17. The relevant [NEP](https://numpy.org/neps/nep-0019-rng-policy.html) says

> We categorically recommend avoiding using the convenience functions when reproducibility is involved.

**Comment by hanlint** [comment]
Thanks @dpaleka for finding this! We were able to reproduce the issue, which was due to an incorrect example in our docstrings. 

If you also seed the `Trainer` with:

```
trainer = composer.Trainer(
    device='gpu',
    model=model,
    train_dataloader=train_dataloader,
    eval_dataloader=eval_dataloader,
    max_duration="5ep",
    algorithms=algorithms,
    seed=42,  # added
) 
```

Then, the results should be reproducible. Across two runs we see:

```
# run 1
Epoch 0: 100%|██████████| 98/98 [00:17<00:00,  5.63it/s, loss/train=2.3508]  

# run 2              
Epoch 0: 100%|██████████| 98/98 [00:17<00:00,  5.59it/s, loss/train=2.3508] 
```

Essentially, if a seed is not passed to the Trainer, we choose a random seed based on system time, leading to the determinism issues above. This is admittedly non-intuitive behavior, but we've now fixed the example and added a section on reproducibility in the guide in #794 .

**Comment by dpaleka** [comment]
Yes, seeding the `Trainer` fixes this issue. Thanks!
